src/blocks/weather.rs

```diff
@@ -0,0 +1,169 @@
+use darksky::{self, Unit, Icon};
+use serde_json::{self, Value};
+use std::boxed::Box;
+use std::error::Error;
+use std::sync::mpsc::Sender;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::time::{Duration, Instant};
+use uuid::Uuid;
+
+use widget::I3BarWidget;
+use widgets::text::TextWidget;
+use scheduler::Task;
+use block::Block;
+
+const DEFAULT_UPDATE_INTERVAL: u64 = 30;
+const DEFAULT_UNIT: InputUnit = InputUnit::Si;
+
+pub struct Weather {
+    id: String,
+    widget: TextWidget,
+    config: Config,
+    forecast: Arc<Mutex<Option<darksky::Forecast>>>
+}
+
+#[derive(Clone, Deserialize)]
+enum InputUnit {
+    #[serde(rename = "SI")]
+    Si,
+    #[serde(rename = "US")]
+    Us
+}
+
+impl InputUnit {
+    fn to_str(&self) -> &'static str {
+        match self {
+            &InputUnit::Si => "℃",
+            &InputUnit::Us => "℉"
+        }
+    }
+}
+
+impl From<InputUnit> for Unit {
+    fn from(unit: InputUnit) -> Unit {
+        match unit {
+            InputUnit::Si => Unit::Si,
+            InputUnit::Us => Unit::Us
+        }
+    }
+}
+
+#[derive(Clone, Deserialize)]
+struct Config {
+    token: String,
+    latitude: f64,
+    longitude: f64,
+    interval: Option<u64>,
+    units: Option<InputUnit>
+}
+
+impl Weather {
+    pub fn new(config: Value, send: Sender<Task>, theme: Value) -> Weather {
+        let id: String = Uuid::new_v4().simple().to_string();
+        let id_clone = id.clone();
+        
+        let config: Config = match serde_json::from_value(config) {
+            Ok(config) => config,
+            Err(error) => panic!("Error in weather block configuration: {}",
+                                 error.description())
+        };
+        let config_clone = config.clone();
+
+        let forecast = Arc::new(Mutex::new(None));
+        let forecast_clone = forecast.clone();
+
+        let widget = TextWidget::new(theme);
+
+        thread::spawn(move || {
+            let id = id_clone;
+            let config = config_clone;
+            let forecast = forecast_clone;
+            let units = config.units.unwrap_or(DEFAULT_UNIT);
+            loop {
+                let set_units = |options: darksky::Options| {
+                    options.unit(units.clone().into())
+                };
+                if let Ok(data) =
+                    darksky::get_forecast_with_options(config.token.clone(),
+                                                       config.latitude,
+                                                       config.longitude,
+                                                       set_units) {
+                    let mut forecast = forecast.lock().unwrap();
+                    *forecast = Some(data);
+                    send.send(Task {
+                        id: id.clone(),
+                        update_time: Instant::now()
+                    }).unwrap();
```

Comment:
I agree this should be done in a separate thread. Usually however I would discourage this pattern if the update() doesn't download information from the internet, because the scheduler can then group updates of blocks that have the same intervals, that way avaoiding redrawing the bar twice in a very short period of time.